fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/metadata/KeyExpressionTest.java

```diff
@@ -725,103 +730,113 @@ public void testSplitConcat() throws Exception {
+                Arguments.of(field("a", FanType.Concatenate),
                         field("a", FanType.None),
-                        false},
-                {field("a", FanType.None),
+                        false),
+                Arguments.of(field("a", FanType.None),
                         field("a", FanType.Concatenate),
-                        false},
-                {field("a", FanType.None),
+                        false),
+                Arguments.of(field("a", FanType.None),
                         field("a", FanType.FanOut),
-                        false},
-                {field("a", FanType.FanOut).nest("b"),
+                        false),
+                Arguments.of(field("a", FanType.FanOut).nest("b"),
                         field("a", FanType.FanOut).nest(concat(field("b"), field("c"))),
-                        true},
-                {field("a", FanType.FanOut).nest("b"),
+                        true),
+                Arguments.of(field("a", FanType.FanOut).nest("b"),
                         concat(field("a", FanType.FanOut).nest("b"), field("a", FanType.FanOut).nest("c")),
-                        true},
-                {field("a", FanType.FanOut).nest(concat(field("b"), field("c"))),
+                        true),
+                Arguments.of(field("a", FanType.FanOut).nest(concat(field("b"), field("c"))),
                         concat(field("a", FanType.FanOut).nest("b"), field("a", FanType.FanOut).nest("c")),
-                        false},
-                {concat(field("a"), VERSION),
+                        false),
+                Arguments.of(concat(field("a"), VERSION),
                         concat(field("a"), field("b")),
-                        false},
-                {concat(field("a"), VERSION),
+                        false),
+                Arguments.of(concat(field("a"), VERSION),
                         concat(field("a"), VERSION),
-                        true},
-                {field("a").split(3),
+                        true),
+                Arguments.of(field("a").split(3),
                         concat(field("a").split(3), field("b"), field("c")),
-                        true},
-                {field("a").split(2),
+                        true),
+                Arguments.of(field("a").split(2),
                         field("a").split(3),
-                        false},
-                {field("a").split(3),
+                        false),
+                Arguments.of(field("a").split(3),
                         field("a").split(2),
-                        false}
-        };
-
-        for (Object[] test : tests) {
-            assertEquals(test[2],
-                    ((KeyExpression) test[0]).isPrefixKey((KeyExpression) test[1]),
-                    "\nPrefix: " + test[0] + "\nKey: " + test[1]);
-        }
+                        false),
+                Arguments.of(keyWithValue(concat(field("a"), field("b")), 1),
```

Comment:
Probably would be beneficial to have a test where `keyWithValue` is the first parameter and the result is `false`, e.g.
```
Arguments.of(keyWithValue(concat(field("a"), field("b"), field("c")), 2),
    concat(field("a"), field("c")), false)
```

Also, for the problematic cases of `keyWithValue(field("a", FanOut).nest(concat(field("b"), field("c"), field("d")), 2)` which should have as prefixes `field("a", FanOut).nest(concat(field("b"), field("c"))` and `field("a", FanOut).nest(concat(field("b"), field("c")))`, but not `field("a", FanOut).nest(concat(field("b"), field("c"), field("d")))` or `concat(field("a", FanOut).nest(field("b")), field("a", FanOut).nest("b"))` Although, if those fail, I don't think it's a regression, but if they pass, then we should be good.

Reply:
All of those work except that the following is false:
```
KeyExpression prefix = field("a", FanOut).nest(concat(field("b"), field("c"));
prefix.isPrefixKey(keyWithValue(field("a", FanOut).nest(concat(field("b"), field("c"), field("d")), 2)) // false
```
When I dig into it, I notice that `prefix.getKeyExpression()` is actually:
```
[Field { 'a' FanOut}/Field { 'b' None}, Field { 'a' FanOut}/Field { 'c' None}]
```
which doesn't look right at all. It seems to be using `normalizeKeyForPositions()` behind the scenes, which has this Javadoc:
```
     * Get key in normalized form for comparing field-by-field.
     * Does not take account of fan-type, so only valid
     * when this has already been taken care of, such as individual index entries.
     * @return a list of key expressions in order
```
which suggests that this is pretty broken. :(

Reply:
This should all be fixed using the new `getKeyExpression()` implementation using `getSubKey()` (see 4238c30).